infrastructure/async/src/main/java/tech/pegasys/teku/infrastructure/async/ThrottlingTaskQueue.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2020 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.infrastructure.async;
+
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Supplier;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.hyperledger.besu.plugin.services.MetricsSystem;
+import tech.pegasys.teku.infrastructure.metrics.SettableGauge;
+import tech.pegasys.teku.infrastructure.metrics.TekuMetricCategory;
+
+public class ThrottlingTaskQueue {
+  private static final Logger LOG = LogManager.getLogger();
+  private final int maximumConcurrentRequests;
+  private volatile Queue<Runnable> queuedRequests = new ConcurrentLinkedQueue<>();
+  private volatile AtomicInteger activeTasks = new AtomicInteger(0);
+
+  private final SettableGauge queueSizeGauge;
+
+  public ThrottlingTaskQueue(
+      final int maximumConcurrentRequests,
+      final MetricsSystem metricsSystem,
+      final TekuMetricCategory metricCategory,
+      final String metricsPrefix) {
+    this.maximumConcurrentRequests = maximumConcurrentRequests;
+
+    this.queueSizeGauge =
+        SettableGauge.create(
+            metricsSystem,
+            metricCategory,
+            metricsPrefix + "_tasks_queued",
+            "Number of tasks queued");
+  }
+
+  public <T> SafeFuture<T> queueRequest(final Supplier<SafeFuture<T>> request) {
+    final SafeFuture<T> future = new SafeFuture<>();
+    queuedRequests.add(
+        () -> {
+          final SafeFuture<T> requestFuture = request.get();
+          requestFuture.propagateTo(future);
+          requestFuture.always(this::requestComplete);
+        });
+    queueSizeGauge.set(queuedRequests.size());
+    processQueuedRequests();
+    return future;
+  }
+
+  private void requestComplete() {
+    activeTasks.decrementAndGet();
+    processQueuedRequests();
+  }
+
+  private void processQueuedRequests() {
```

Comment:
@ajsutton processQueuedResults and requestComplete were synchronized in their original spot.
I've added an atomic integer to act as the limit counter, and made them not synchronized, but I'll be interested to get your take on the best way to go here...

Reply:
The `synchronized` was essential and you'll need it back. :). The first key race condition here is that there's a race between when you check `queuedRequests.isEmpty()` and `queuedRequests.remove().run()` - the latter will give you either a `NoSuchElementException` or `NullPointerException` if the last item in the queue is removed by a different thread between the two calls.

The next race condition you'll hit once you fix that is that after you call `incrementAndGet` and find that it's actually too high, some tasks may complete which will then decrement but still not make any new requests because the limit is still reached.  Then you'll decrement to below the limit but not make any new requests and we're left underutilising the limit.

A throttling queue is actually incredibly complex to get right with a lock free approach and there's actually no performance benefit to avoiding locks anyway.